File: src/parsers/javascript.py
from .base import BaseParser


def _preceding_decorators(node, text_fn) -> list[str]:
    """收集紧邻该节点之前的 decorator 兄弟节点"""
    if not node.parent:
        return []
    decorators = []
    for child in node.parent.children:
        if child == node:
            break
        if child.type == "decorator":
            decorators.append(text_fn(child))
        else:
            decorators.clear()  # 只取紧邻的，中间有其他节点就重置
    return decorators


class JavaScriptParser(BaseParser):
    LANG = "javascript"

    QUERY = """
        [
          (function_declaration name: (identifier) @name)
          (method_definition name: (property_identifier) @name)
        ] @method
    """

    CLASSES_QUERY = """
        (class_declaration name: (identifier) @name) @class
    """

    CALLS_QUERY = """
        (call_expression function: [
          (identifier) @call
          (member_expression property: (property_identifier) @call)
        ])
    """

    CLASS_NODE_TYPES = ["class_declaration"]

    def _build_signature(self, node) -> str:
        name = node.child_by_field_name("name")
        params = node.child_by_field_name("parameters")
        sig = (self._text(name) if name else "") + (self._text(params) if params else "()")
        return sig


class TypeScriptParser(BaseParser):
    LANG = "typescript"

    QUERY = """
        [
          (function_declaration name: (identifier) @name)
          (method_definition name: (property_identifier) @name)
        ] @method
    """

    CLASSES_QUERY = """
        (class_declaration name: (type_identifier) @name) @class
    """

    CALLS_QUERY = """
        (call_expression function: [
          (identifier) @call
          (member_expression property: (property_identifier) @call)
        ])
    """

    CLASS_NODE_TYPES = ["class_declaration"]

    def _get_annotations(self, node) -> list[str]:
        return _preceding_decorators(node, self._text)

    def _build_signature(self, node) -> str:
        name = node.child_by_field_name("name")
        params = node.child_by_field_name("parameters")
        return_type = node.child_by_field_name("return_type")

        sig = (self._text(name) if name else "") + (self._text(params) if params else "()")
        if return_type:
            sig += f": {self._text(return_type)}"

        decorators = self._get_annotations(node)
        if decorators:
            return " ".join(decorators) + " " + sig
        return sig

File: src/parser/javascript.rs
use anyhow::Result;
use tree_sitter::{Language, Node};

use super::base::{CoreParser, LanguageParser};

const QUERY: &str = r#"
    [
      (function_declaration name: (identifier) @name) @method
      (method_definition name: (property_identifier) @name) @method
    ]
"#;

const CLASSES_QUERY: &str = r#"
    (class_declaration name: (identifier) @name) @class
"#;

pub struct JavaScriptParser(pub CoreParser);

impl JavaScriptParser {
    pub fn new(source: Vec<u8>) -> Result<Self> {
        let language = Language::from(tree_sitter_javascript::LANGUAGE);
        Ok(Self(CoreParser::new(
            source,
            language,
            QUERY,
            Some(CLASSES_QUERY),
            None,
        )?))
    }
}

impl LanguageParser for JavaScriptParser {
    fn core(&self) -> &CoreParser {
        &self.0
    }

    fn class_node_types(&self) -> &[&str] {
        &["class_declaration", "class"]
    }

    fn build_signature(&self, node: Node) -> String {
        let core = &self.0;
        let name = node.child_by_field_name("name").map(|n| core.text(n)).unwrap_or("");
        let params = node
            .child_by_field_name("parameters")
            .map(|n| core.text(n))
            .unwrap_or("()");
        format!("{}{}", name, params)
    }
}
